hudi-spark-datasource/hudi-spark2/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/Spark24ParquetReader.scala

```diff
@@ -156,30 +177,51 @@ class Spark24ParquetReader(enableVectorizedReader: Boolean,
       iter.asInstanceOf[Iterator[InternalRow]]
     } else {
       // ParquetRecordReader returns UnsafeRow
+      val readSupport = new ParquetReadSupport(convertTz)
       val reader = if (pushed.isDefined && enableRecordFilter) {
         val parquetFilter = FilterCompat.get(pushed.get, null)
-        new ParquetRecordReader[UnsafeRow](new ParquetReadSupport(convertTz), parquetFilter)
+        new ParquetRecordReader[UnsafeRow](readSupport, parquetFilter)
       } else {
-        new ParquetRecordReader[UnsafeRow](new ParquetReadSupport(convertTz))
+        new ParquetRecordReader[UnsafeRow](readSupport)
       }
       val iter = new RecordReaderIterator(reader)
       // SPARK-23457 Register a task completion lister before `initialization`.
       taskContext.foreach(_.addTaskCompletionListener[Unit](_ => iter.close()))
       reader.initialize(split, hadoopAttemptContext)
 
       val fullSchema = requiredSchema.toAttributes ++ partitionSchema.toAttributes
-      val joinedRow = new JoinedRow()
-      val appendPartitionColumns = GenerateUnsafeProjection.generate(fullSchema, fullSchema)
+      val unsafeProjection = if (implicitTypeChangeInfos.isEmpty) {
+        GenerateUnsafeProjection.generate(fullSchema, fullSchema)
+      } else {
+        val newFullSchema = new StructType(requiredSchema.fields.zipWithIndex.map { case (f, i) =>
+          if (implicitTypeChangeInfos.containsKey(i)) {
+            StructField(f.name, implicitTypeChangeInfos.get(i).getRight, f.nullable, f.metadata)
+          } else f
+        }).toAttributes ++ partitionSchema.toAttributes
+        val castSchema = newFullSchema.zipWithIndex.map { case (attr, i) =>
+          if (implicitTypeChangeInfos.containsKey(i)) {
+            val srcType = implicitTypeChangeInfos.get(i).getRight
+            val dstType = implicitTypeChangeInfos.get(i).getLeft
+            val needTimeZone = Cast.needsTimeZone(srcType, dstType)
+            Cast(attr, dstType, if (needTimeZone) timeZoneId else None)
+          } else attr
+        }
+        GenerateUnsafeProjection.generate(castSchema, newFullSchema)
+      }
```

Comment:
This part can still use `schemaEvolutionUtils.generateUnsafeProjection`?

Reply:
Spark 2 schema evolution diverges, so we just do what's in the legacy format. We can follow up and discuss with the original authors of schema on read